### PR TITLE
Libnetwork to validate Ipam returned addresses

### LINKED ISSUE #655: 1.9.0-rc1 : libnetwork must validate IPAM drivers responses
If the IPAM driver doesn't return an IP/IPv6 address/pool with a nil error, it is an error condition and libnetwork core should not allow it.

### PR DESCRIPTION
**Pull request by aboch**
Closes #655 

Signed-off-by: Alessandro Boch <aboch@docker.com>

### CODE REVIEW

File: endpoint.go [hunk 1]
```diff
@@ -638,7 +638,24 @@ func (ep *endpoint) assignAddress() error {
...
 	}
-	return ep.assignAddressVersion(6, ipam)
+
+	err = ep.assignAddressVersion(6, ipam)
+	if err != nil {
+		return err
+	}
+
+	// Verify Ipam driver returned valid IP addresses for this endpoint
+	list := []*net.IPNet{ep.Iface().Address()}
```
Comment (mrjana) [inline endpoint.go]: Why don't you just check if `ep.Iface().Address() == nil || (len(n.ipamV6Info != 0 && ep.Iface().AddressIPv6() == nil`?

Comment (aboch) [inline endpoint.go]: I guess the answer is over engineering ;-). 
Let me fix it following your suggestion.
Thanks

**Comment by aboch** [comment]
@mrjana Taken care of your comment. Also updated the ipam doc accordingly. 
PTAL when you get a chance.

**Comment by mrjana** [comment]
LGTM

**Comment by mavenugo** [comment]
@aboch @mrjana I understand the intention of this PR. But, I think we should hold on this till we understand the use-cases better. Maybe it is better to let the user decide this instead of hard-coding such a logic in libnetwork core. WDYT ?

**Comment by aboch** [comment]
@mavenugo 
I am fine to remove this restriction. As we discussed offline, we might want to support a `null` ipam driver option in future.

Please close #655 then.

I am ok to close this PR now.

**Comment by pyakpyak** [comment]
Thanks @aboch @mavenugo 

+1 for less restrictive behavior and null IPAM support. In our use case (assigning secondary addresses from ENI in VPC+L3 routing on host) it seems pretty tricky to separate out the logic between two drivers.

**Pull request closed without merging**